numpy/core/src/multiarray/methods.c

```diff
@@ -1634,6 +1635,98 @@ array_reduce(PyArrayObject *self, PyObject *NPY_UNUSED(args))
     return ret;
 }
 
+static PyObject *
+array_reduce_ex(PyArrayObject *self, PyObject *args)
+{
+    int protocol;
+    PyObject *ret = NULL, *numeric_mod = NULL, *from_buffer_func = NULL;
+    PyObject *buffer_tuple = NULL, *pickle_module = NULL, *pickle_class = NULL;
+    PyObject *class_args = NULL, *class_args_tuple = NULL, *unused = NULL;
+    PyObject *subclass_array_reduce = NULL;
+    PyObject *buffer = NULL;
+    PyArray_Descr *descr = NULL;
+
+    if (PyArg_ParseTuple(args,"i", &protocol)){
+        descr = PyArray_DESCR(self);
+        if ((protocol<5) ||
+            !PyArray_IS_C_CONTIGUOUS((PyArrayObject*)self) ||
+            PyDataType_FLAGCHK(descr, NPY_ITEM_HASOBJECT) ||
+            PyDataType_ISUNSIZED(descr)) {
+            /* The PickleBuffer class from version 5 of the pickle protocol
+             * can only be used for arrays backed by a contiguous data buffer.
+             * For all other cases we fallback to the generic array_reduce
+             * method that involves using a temporary bytes allocation. However
+             * we do not call array_reduce directly but instead lookup and call
+             * the __reduce__ method to make sure that it's possible customize
+             * pickling in sub-classes. */
```

Comment:
It would also be great to also add protocol 5 support for arrays that have the F-contiguous flag but not the C-contiguous flag: in this case we should reduce self.T and then take the transpose back when reconstructing:

```python
 _from_buffer(buffer, dtype, shape, transpose=True)
```